### PR TITLE
Misc fixes to Conv and ConvTranspose CUDA kernels

### LINKED ISSUE #4014: The output is different with the same input on CUDA
**Describe the bug**
I trained a MaskRcnn model and convert it to onnx. When i use test images to do inference,it can give the correct output. But if i do the inference with a white image (all 255) first, and then use test images to do inference, it can't recognize correctly with onnxruntime-gpu1.3.0. 
if i use onnxruntime1.3.0, it will not happen.

**System information**
- OS Platform and Distribution :Windows 10 (c++) & Linux Ubuntu 16.04(python3.7)
- ONNX Runtime installed from :binary from pip install
- ONNX Runtime version:1.3.0
- Python version:3.7
- Visual Studio version (if applicable):2017
- GCC/Compiler version (if compiling from source):compiling from binary
- CUDA/cuDNN version:CUDA10.1 & cuDNN7.6.5
- GPU model and memory:2080Ti 11G

**To Reproduce**
- Describe steps/code to reproduce the behavior.
- Attach the ONNX model to the issue (where applicable) to expedite investigation.
onnx model and images are here [https://drive.google.com/open?id=1c7GDq8AYgneP6mdqFPSH1GKl7Oggn8Mq](url)
do the inference with a white image (all 255) first, and then use test images to do inference
```
import os
import onnxruntime
import cv2
import torch #1.5.0
from detectron2.data.transforms as T #just for resize
filelist = os.listdir('./')
sess = onnxruntime.InferenceSession('./segmentation.onnx')
for i in filelist:
    inputs = cv2.imread('./' + i)
    transform_gen = T.ResizeShortestEdge([800,800],1333)
    inputs = transform_gen.get_transform(inputs).apply_image(inputs)
    inputs = torch.as_tensor(inputs.astype('float32').transpose(2,0,1))
    inputs = inputs.unsqueeze(0)
    res = sess.run(None, {sess.get_inputs()[0].name:inputs.numpy()})
    print (res)
```

**Screenshots**
Screenshot from 2020-05-21 16-50-27

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: There are some subtle fixes to the `Conv` and `ConvTranspose` CUDA kernels:

1) In ConvTranspose kernel, extend holding onto the mutex until we are done with all CuDNN calls (Similar to #3837 which did that for the Conv CUDA kernel)

2) Fixes corresponding to bailing out too early in the Compute methods if one of the output dim values is 0- 

ConvTranpose kernel:
a) There is logic to bail out early if one of the output dimensions is zero, but this logic resides in an `if` block that only executes for the first run or on a run whose input shape/ filter shape is different from the cached input shape / filter shape from a previous run which meant if the output was empty on one run and the input shape / filter shape remains unchanged on the next run, the logic to bail out early is never hit and will make the CuDNN call resulting in a runtime failure. So fix this.

b) If bailing out early in the very first run, cache the `output shape` and `filter dimensions` for use in subsequent runs

Conv kernel:

a) If bailing out early in the very first run, cache the `filter dimensions` for use in subsequent runs

**Motivation and Context**
Fix #4014

### CODE REVIEW

File: onnxruntime/core/providers/cuda/nn/conv_transpose.cc [hunk 1]
```diff
@@ -155,42 +157,49 @@ Status ConvTranspose<T>::DoConvTranspose(OpKernelContext* context, bool dynamic_
...
+            &beta,
+            s_.y_tensor,
+            y_data));
+
+    if (has_bias) {
+      const Tensor* B = dynamic_padding ? context->Input<Tensor>(3) : context->Input<Tensor>(2);
+      auto b_data = reinterpret_cast<const CudaT*>(B->template Data<T>());
+      CUDNN_RETURN_IF_ERROR(cudnnAddTensor(CudnnHandle(), &alpha, s_.b_tensor, b_data, &alpha, s_.y_tensor, y_data));
+    }
   }
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/nn/conv_transpose.cc:203]: Move this brace down thus extending holding onto the mutex after all CuDNN calls

File: onnxruntime/core/providers/cuda/nn/conv_transpose.cc [hunk 2]
```diff
@@ -82,11 +82,6 @@ Status ConvTranspose<T>::DoConvTranspose(OpKernelContext* context, bool dynamic_
       ConvTransposeAttributes::Prepare p;
       ORT_RETURN_IF_ERROR(conv_transpose_attrs_.PrepareForCompute(context, has_bias, p, dynamic_padding));
 
-      // Bail out early if one of the dimensions is zero.
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/nn/conv_transpose.cc:85]: Move this block until after `s_.y_dims` has been cached for future runs

File: onnxruntime/core/providers/cuda/nn/conv_transpose.cc [hunk 3]
```diff
@@ -155,42 +155,49 @@ Status ConvTranspose<T>::DoConvTranspose(OpKernelContext* context, bool dynamic_
       s_.algo = perf.algo;
       s_.workspace_bytes = perf.memory;
     }
-  }
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/nn/conv_transpose.cc:158]: Move this brace down to extend holding onto the mutex longer

File: onnxruntime/core/providers/cuda/nn/conv_transpose.cc [hunk 4]
```diff
@@ -155,42 +155,49 @@ Status ConvTranspose<T>::DoConvTranspose(OpKernelContext* context, bool dynamic_
...
+    if (!y_data) {
+      auto y_dims = s_.y_dims;
+      if (x_dimensions == 3) {
+        y_dims.erase(y_dims.begin() + 2);
+      }
+      Tensor* Y = context->Output(0, TensorShape(y_dims));
+      y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T>());
+
+      // Bail out early if one of the output dimensions is zero.
+      if (Y->Shape().Size() == 0) {
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/nn/conv_transpose.cc:170]: `if` logic to bail out early

File: onnxruntime/core/providers/cuda/nn/conv_transpose.cc [hunk 5]
```diff
@@ -98,6 +93,11 @@ Status ConvTranspose<T>::DoConvTranspose(OpKernelContext* context, bool dynamic_
       }
       s_.y_dims = y_dims;
 
+      // Bail out early if one of the dimensions is zero.
+      if (p.Y->Shape().Size() == 0) {
+        return Status::OK();
+      }
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/nn/conv_transpose.cc:100]: Would be good to capture in this comment that we need to wait until s_.y_dims is set so it doesn't get moved back accidentally in the future.

Comment (hariharans29) [reply onnxruntime/core/providers/cuda/nn/conv_transpose.cc:100]: sure, I ll add the comment. I have one more commit to push to truly resolve #4014

Comment (hariharans29) [reply onnxruntime/core/providers/cuda/nn/conv_transpose.cc:106]: Added the comment
